File: native/Travel Companion/Model/PlannableFactory.swift
//
//  PlannableFactory.swift
//  Travel Companion
//
//  Created by Stefan Jaindl on 12.09.18.
//  Copyright © 2018 Stefan Jaindl. All rights reserved.
//

import CodableFirebase
import Foundation
import shared

class PlannableFactory {
    static func makePlannable(of type: String, data: [String: Any]) throws -> Plannable {
      switch type {
        case Constants.Plannables.flight:
            return try FirestoreDecoder().decode(Flight.self, from: data)
        case Constants.Plannables.publicTransport:
            return try FirestoreDecoder().decode(PublicTransport.self, from: data)
        case Constants.Plannables.attraction:
            return try FirestoreDecoder().decode(GooglePlace.self, from: data)
        case Constants.Plannables.hotel:
            return try FirestoreDecoder().decode(GooglePlace.self, from: data)
        case Constants.Plannables.restaurant:
            return try FirestoreDecoder().decode(GooglePlace.self, from: data)
         
        default:
            throw NSError(domain: "Plannable type not supported", code: -1, userInfo: [:])
        }
    }
}

File: app/src/main/java/com/sjaindl/travelcompanion/plan/PlannableFactory.kt
package com.sjaindl.travelcompanion.plan

import com.sjaindl.travelcompanion.Constants
import com.sjaindl.travelcompanion.api.google.GooglePlace
import com.sjaindl.travelcompanion.api.google.Photo
import com.sjaindl.travelcompanion.api.google.Plannable
import com.sjaindl.travelcompanion.api.google.asPlannable
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.properties.Properties
import kotlinx.serialization.properties.decodeFromMap

class PlannableFactory {
    companion object {
        @OptIn(ExperimentalSerializationApi::class)
        fun makePlannable(type: String, data: Map<String, Any>): Plannable {
            val place: GooglePlace = when (type) {
                Constants.Plannables.attraction -> {
                    Properties.decodeFromMap(data)
                }

                Constants.Plannables.hotel -> Properties.decodeFromMap(data)
                Constants.Plannables.restaurant -> Properties.decodeFromMap(data)
                else -> throw Exception("Plannable type not supported")
            }

            // Properties.decodeFromMap only supports flat types so far..
            val photoList = data["photos"] as? List<Map<String, Any>>?
            place.photos = photoList?.mapNotNull {
                Photo.customDecode(photos = it)
            } ?: emptyList()

            place.htmlAttributions = data["html_attributions"] as? List<String>
            place.types = data["types"] as? List<String>

            return place.asPlannable()
        }
    }
}
